### PR TITLE
OpenAI Mock Backend

### PR DESCRIPTION
**Pull request by TimPansino**
# Overview

* Add mocked backend for OpenAI testing.
* Allow recording of requests to OpenAI to and audit log for mock updating.

### CODE REVIEW

File: tests/testing_support/mock_external_openai_server.py [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+#
+# 1) This removes dependencies on external websites.
+# 2) Provides a better mechanism for making an external call in a test app than
+#    simple calling another endpoint the test app makes available because this
+#    server will not be instrumented meaning we don't have to sort through
+#    transactions to separate the ones created in the test app and the ones
+#    created by an external call.
+# 3) This app runs on a separate thread meaning it won't block the test app.
+
+ALLOWED_PATHS = set(["/chat/completions"])
```
Comment (umaannamalai) [inline tests/testing_support/mock_external_openai_server.py:31]: We will also need to test a fake path for errors and the embeddings endpoint as well. Could that be added to this?

Comment (TimPansino) [reply tests/testing_support/mock_external_openai_server.py:31]: Honestly could probably drop this check altogether, it doesn't feel as necessary after I used the SDK enough.

Comment (umaannamalai) [reply tests/testing_support/mock_external_openai_server.py:31]: That's fair - we just need a mock response in here for the embeddings response object

Comment (TimPansino) [reply tests/testing_support/mock_external_openai_server.py]: I've updated the embedding tests as well and ensured they work with the mocking setup.

**Review by umaannamalai — APPROVED** [review]

**Review by hmstepanek — APPROVED** [review]
LGTM.

**Review by umaannamalai — APPROVED** [review]

**Pull request merged**